openqa-trigger-bisect-jobs

```diff
@@ -47,28 +48,60 @@ def call(cmds, dry_run=False):
 def openqa_clone(cmds, dry_run, default_opts=['--skip-chained-deps', '--within-instance'], default_cmds=['_GROUP=0']):
     return call(['openqa-clone-job'] + default_opts + cmds + default_cmds, dry_run)
 
+def fetch_url(url, request_type='text'):
+    if url.scheme == "file":
+        decodejson = lambda c: json.loads(c)
+        path = url.geturl()
+        path = path[7:]
+        path = 'test/data/python-requests/' + path
+        with open(path, "r") as request:
+            raw = content = request.read()
+    elif url.scheme in ["http", "https"]:
+        decodejson = lambda c: content.json()
+        try:
+            content = requests.get(url.geturl())
+        except requests.exceptions.RequestException as e:
+            log.error("Error while fetching "+url.geturl()+": "+str(e))
+            raise(e)
+        content.raise_for_status()
+        raw = r.content
+    if request_type == 'json':
+        try:
+            content = decodejson(content)
+        except json.decoder.JSONDecodeError as e:
+            log.error("Error while decoding JSON from "+url.geturl()+" -> >>"+raw+"<<: "+str(e))
+            raise(e)
+    return content
 
 def main():
     args = parse_args()
-    investigation_url = urljoin(args.url + '/', 'investigation_ajax')
-    log.debug('Retrieving investigation info from %s' % investigation_url)
-    out = requests.get(investigation_url)
-    log.debug('Received investigation info: %s' % out)
-    investigation = out.json()
-    os_test_issues = [line for line in investigation['diff_to_last_good'].split('\n') if 'OS_TEST_ISSUES' in line]
-    good, bad = [set(re.compile(': "([^"]*)",').search(line).group(1).split(',')) for line in os_test_issues]
+    parsed_url = urlparse(args.url)
+    investigation_url = parsed_url._replace(path=parsed_url.path + '/investigation_ajax')
+    log.debug('Retrieving investigation info from %s' % investigation_url.geturl())
+    investigation = fetch_url(investigation_url, request_type='json')
+    log.debug('Received investigation info: %s' % investigation)
+    good = bad = set()
+    for line in investigation['diff_to_last_good'].split('\n'):
+        search = re.compile('([+-]) +"OS_TEST_ISSUES" *: "([^"]*)",').search(line)
+        if search:
+            issues = set(search.group(2).split(','))
+            if search.group(1) == '+':
+                bad = issues
+            else:
+                good = issues
```

Comment:
I would try to avoid this procedural code. Another aspect is that it's less efficient although that's not a strong concern here. Why is it necessary to change this? What problematic cases did you encounter?

Reply:
Well, the previous code was failing awfully uninformative when the regex didn't match. `OS_TEST_ISSUES` can be part of another setting, who knows? And then we will have these error messages that there is no attribute `group` in the gru log.

If you think the script is fine as it is, then I will unassign myself from the progress ticket again.

Reply:
replaced with a dictionary